### PR TITLE
irmin-pack: Add sequential accessors & clean dispatcher

### PR DESCRIPTION
**Pull request by clecat**
This PR adds:
- 2 accessors that uses physical offsets instead of the virtual offsets
- A generator for sequences of accessors that iterates trough all physical offsets of a pack file.
It also removes:
- The `root` filed of `Dispatcher.t`, as it is never used.

### CODE REVIEW

File: src/irmin-pack/unix/dispatcher_intf.ml [hunk 1]
```diff
@@ -50,6 +50,38 @@ module type S = sig
   (** [shrink_accessor_exn a ~new_len] is [a] where the length is smaller than
       in [a].*)
 
+  module Sequential : sig
+    val create_accessor_exn : int63 -> int63 -> off:int63 -> len:int -> accessor
```
Comment (clecat) [inline src/irmin-pack/unix/dispatcher_intf.ml:54]: The prefix & suffix length are functions that are very costly (or at least for the prefix as it is a syscall), therefore I took the choice to give those directly to the accessors creators.

Maybe a cleaner solution would be to save those in the `Dispatcher.t`, however it does mean that we would calculate them (and have to handle the potential errors) even when the user will never use the accessors that needs them.

Dont hesitate if you have a better solution

File: src/irmin-pack/unix/dispatcher.ml [hunk 2]
```diff
@@ -250,4 +250,77 @@ module Make (Fm : File_manager.S with module Io = Io.Unix) :
...
+      let prefix_len =
+        match Fm.prefix t.fm with
+        | Some prefix -> (
+            match Io.read_size prefix with
+            | Ok len -> len
+            | Error _ -> Int63.zero)
+        | None -> Int63.zero
+      in
+      let suffix_len = Fm.Suffix.end_offset (Fm.suffix t.fm) in
+      let end_offset = prefix_len + suffix_len in
```
Comment (Ngoguey42) [inline src/irmin-pack/unix/dispatcher.ml:310]: You've introduced a third kind of offsets. I suggest you name these differently in the code, for example `soff`.

At the top of the module `Seq` you could then add this piece of documentation such as this one: `An soff is a sequential offset that indexes inside the virtual concatenation of the prefix and suffix files.`

File: src/irmin-pack/unix/dispatcher.ml [hunk 3]
```diff
@@ -250,4 +250,77 @@ module Make (Fm : File_manager.S with module Io = Io.Unix) :
...
+        let entry_end_offset = off + Int63.of_int len in
+        if entry_end_offset > prefix_len then
+          raise (Errors.Pack_error `Read_out_of_bounds)
+        else { poff = off; len; location = Prefix }
+
+    let create_accessor_from_range_exn prefix_len suffix_len ~off ~min_len
+        ~max_len =
+      let open Int63.Syntax in
+      if off >= prefix_len then
+        let off = off - prefix_len in
```
Comment (Ngoguey42) [inline src/irmin-pack/unix/dispatcher.ml:273]: You could then have such parts modified to make explicit the offset flavours.
```suggestion
      if soff >= prefix_len then
        let poff = soff - prefix_len in
```

File: src/irmin-pack/unix/dispatcher_intf.ml (hunk 1 above)
Comment (Ngoguey42) [reply src/irmin-pack/unix/dispatcher_intf.ml:54]: It makes sense to avoid recomputing the prefix length over and over. 

However, is there any real use for exposing `create_accessor_exn` and `create_accessor_from_range_exn` in the `Dispatcher` signature? Since they introduce a third kind of offset it could be great if that concept stayed hidden inside the implementation of `Sequential` (you could preserve your docstrings to the `.ml` files).

File: src/irmin-pack/unix/dispatcher.ml [hunk 4]
```diff
@@ -250,4 +250,77 @@ module Make (Fm : File_manager.S with module Io = Io.Unix) :
...
+        if off < end_offset then (
+          let accessor =
+            create_accessor_from_range_exn prefix_len suffix_len ~off
+              ~min_len:min_header_len ~max_len:max_header_len
+          in
+          read_exn t accessor buf;
+          let len = read_len buf in
+          Some
+            ( (off, create_accessor_exn prefix_len suffix_len ~off ~len),
+              Int63.(add off (of_int len)) ))
```
Comment (Ngoguey42) [inline src/irmin-pack/unix/dispatcher.ml:321]: I believe that the offset you are returning is an `soff` while what's needed is a global offset

File: src/irmin-pack/unix/dispatcher_intf.ml (hunk 1 above)
Comment (clecat) [reply src/irmin-pack/unix/dispatcher_intf.ml:54]: At first I had them hidden, but for the purpose of [#1939](https://github.com/mirage/irmin/pull/1939), I'll also need direct access using the offset returned

File: src/irmin-pack/unix/dispatcher.ml (hunk 2 above)
Comment (clecat) [reply src/irmin-pack/unix/dispatcher.ml:310]: Good idea, I didnt know how to name it, `soff` sounds good

File: src/irmin-pack/unix/dispatcher.ml (hunk 4 above)
Comment (clecat) [reply src/irmin-pack/unix/dispatcher.ml:321]: It is indeed a `soff` that I am returning, however, isnt it the only offset that is available to me ? If I were to be able to use another one, I could avoid having the 2 accessors public, but I dont think I can easily go from a `soff` to a `global offset` without using `Mapping_file` ?

Comment (Ngoguey42) [reply src/irmin-pack/unix/dispatcher.ml:321]: I think you should traverse the mapping file chunks in order to be able to retrieve the global offsets

File: src/irmin-pack/unix/dispatcher.ml [hunk 5]
```diff
@@ -250,4 +250,80 @@ module Make (Fm : File_manager.S with module Io = Io.Unix) :
...
+      else rem_len
+    in
+    { poff; len; location }
+
+  let create_sequential_accessor_seq t ~min_header_len ~max_header_len ~read_len
+      =
+    let chunks =
+      match Fm.mapping t.fm with
+      | Some mapping ->
+          let chunks = ref [] in
```
Comment (metanivek) [inline src/irmin-pack/unix/dispatcher.ml:272]: Perhaps we name this variable something else in light of upcoming suffix chunks?

File: src/irmin-pack/unix/dispatcher.ml [hunk 6]
```diff
@@ -250,4 +250,80 @@ module Make (Fm : File_manager.S with module Io = Io.Unix) :
...
+          Mapping_file.iter mapping (fun ~off ~len ->
+              chunks := (off, len) :: !chunks)
+          |> Errs.raise_if_error;
+          List.rev !chunks
+      | None -> []
+    in
+    let suffix_end_offset = Fm.Suffix.end_offset (Fm.suffix t.fm) in
+    let entry_offset_suffix_start = entry_offset_suffix_start t in
+    let buf = Bytes.create max_header_len in
```
Comment (metanivek) [inline src/irmin-pack/unix/dispatcher.ml:281]: I haven't thought about the full implications here, but since we use `Bytes.unsafe_to_string` quite a bit in the code, I think we ought to at least consider if using a shared buffer here is appropriate.

File: src/irmin-pack/unix/dispatcher_intf.ml [hunk 7]
```diff
@@ -50,6 +50,20 @@ module type S = sig
...
+    max_header_len:int ->
+    read_len:(bytes -> int) ->
+    (int63 * accessor) Seq.t
+  (** [create_sequential_accessor_seq ~min_header_len ~max_header_len ~read_len]
+      returns a sequence of accessors, which simulates iterating sequentially
+      trough the entries of a pack file. [min_header_len] & [max_header_len]
+      represents the minimum & maximum lengths required to read the header of an
+      entry. [read_len] will then be called with a buffer containing the header
+      of the entry and should return the total length of the entry (the length
+      of he header plus the length of the payload)*)
```
Comment (metanivek) [inline src/irmin-pack/unix/dispatcher_intf.ml:65]: Could you add some tests that demonstrate the usage of this new api? In addition to testing the code, it would help me think more about this particular api.

File: src/irmin-pack/unix/dispatcher.ml [hunk 8]
```diff
@@ -250,4 +250,80 @@ module Make (Fm : File_manager.S with module Io = Io.Unix) :
...
+              create_sequential_accessor_from_range_exn Prefix rem_len ~poff
+                ~min_len:min_header_len ~max_len:max_header_len
+            in
+            read_exn t accessor buf;
+            let entry_len = read_len buf in
+            let r =
+              ( off,
+                create_sequential_accessor_exn Prefix rem_len ~poff
+                  ~len:entry_len )
+            in
```
Comment (metanivek) [inline src/irmin-pack/unix/dispatcher.ml:320]: Might be nice to factor this common code pattern (it is essentially the same for suffix and prefix) into one function.

File: src/irmin-pack/unix/dispatcher.ml (hunk 5 above)
Comment (clecat) [reply src/irmin-pack/unix/dispatcher.ml:272]: Will prefix_blocks or even prefix_chunks do the trick ?

File: src/irmin-pack/unix/dispatcher.ml (hunk 6 above)
Comment (clecat) [reply src/irmin-pack/unix/dispatcher.ml:281]: This byte is a bit different from the others, as we know from the start which size will always fit (max_header_len, which is a small size as well).

We could consider using the same `Bytes.t` as other places, but I do not know if we will really gain much from it

File: src/irmin-pack/unix/dispatcher_intf.ml (hunk 7 above)
Comment (clecat) [reply src/irmin-pack/unix/dispatcher_intf.ml:65]: My bad, it is meant to be later used in #1939, it is part of the process of getting rid of duplicate code in it.
On the matter of tests however, I have never created any for irmin, so I might not be the best person for them

File: src/irmin-pack/unix/dispatcher.ml (hunk 6 above)
Comment (metanivek) [reply src/irmin-pack/unix/dispatcher.ml:281]: Ah, my bad. My last sentence is ambiguous. Let me try again. :sweat_smile: 

`buf` here is reused across multiple invocations of (now) `get_entry_accessor` so it is shared. It likely is not a problem with how you want to use it (if `read_len` promises to not do anything with `Bytes.unsafe_to_string` and store/use the results after its initial invocation), but it would (I think) be safer to create a new `buf` for each invocation of `get_entry_accessor`.

Comment (clecat) [reply src/irmin-pack/unix/dispatcher.ml:281]: I see, a better solution would be to give `read_len` a string and use `Bytes.unsafe_to_string` ourselves. It is what was planned at the beginning when @Ngoguey42 and me talked about it, but I forgot to do it.

This way we have no issues with the sub-function modifying it (except if they do some shenanigans with `unsafe_of_string`, which is highly unlikely to happen) and we do not recreate the buffer every time.

Is it a good compromise or should we really go for an initialization every times ?

File: src/irmin-pack/unix/dispatcher_intf.ml (hunk 7 above)
Comment (clecat) [reply src/irmin-pack/unix/dispatcher_intf.ml:65]: I was looking it up and indeed, it does not seem that the dispatcher is yet directly tested. If we don't mind, we should go for what you proposed but keep in mind that tests will be necessary in a close future, maybe via an issue

Comment (metanivek) [reply src/irmin-pack/unix/dispatcher_intf.ml:65]: I agree that it would be nice to have tests for lower levels of the stack -- thanks for opening https://github.com/mirage/irmin/issues/2092. I think it is okay to proceed as-is; perhaps we can pick up this testing as a part of our upcoming GC work.

File: src/irmin-pack/unix/dispatcher.ml (hunk 6 above)
Comment (metanivek) [reply src/irmin-pack/unix/dispatcher.ml:281]: So something like?

```ocaml
let str = Bytes.unsafe_to_string buf in
read_len str
```

The issue with this is if `read_len` stores `str` and tries to use it later (I have no reason to think it would but merely thinking through possibilities). Once `buf` is overwritten by a subsequent accessor, the original `str` will be different since it is using `buf` as its storage.

An excerpt from [docs on `unsafe_to_string`](https://v2.ocaml.org/api/Bytes.html#unsafe):

> Unique ownership is linear: passing the data to another piece of code means giving up ownership (we cannot write the data again). A unique owner may decide to make the data shared (giving up mutation rights on it), but shared data may not become uniquely-owned again.

Without in-depth thinking, I do like handing `read_len` a string. If we make that change, I see two options to be safe:

1. Do not re-use `buf` and use `Bytes.unsafe_to_string` (unique ownership is passed along)
2. Re-use `buf` and use `Bytes.to_string` (no need to think about unsafe operations)

Comment (clecat) [reply src/irmin-pack/unix/dispatcher.ml:281]: Good point, I would go for the first option

**Review by Ngoguey42 — APPROVED** [review]
Thanks. Ship it
